Add neighborhood detail modal

rachelekm opened the pull request:
## Overview

Adds to following to implement the neighborhood details modal in line with figma designs:
- Adds interactivity to update URL path and open neighborhood details on list card selection
- Adds responsive neighborhood detail header, info/units toggle in line with figma designs
- Adds all info tab content with exception of "Your Trips" (to be added in #719 )
- Adds static transit map header
- Adds units tab content
- Adds units site selection modal
- Adds translations setup for all neighborhood details content

### Checklist

- [x] `fixup!` commits have been squashed
- [x] `CHANGELOG.md` updated with summary of features or fixes, following
      [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [ ] `README.md` updated if necessary to reflect the changes
- [x] Run `./scripts/format` to lint, format, and fix the application source code.
- [x] CI passes after rebase

### Demo

https://github.com/user-attachments/assets/377d5716-5df9-424b-9a84-3a8eea07c14a


## Testing Instructions

 * `scripts/server`
 * Set screen to mobile-width and open the list view of the discover page
 * Click on any neighborhood card
 * Confirm neighborhood detail modal opens and URL path has updated to include zipcode
 * Exit modal and confirm path resets to previous location
 * Click on a neighborhood again and confirm modal re-opens and URL path has updated again
 * Confirm static map at top of neighborhood detail with correct neighborhood bounds
 * Confirm static map shows MTA subway lines as well as bus lines and stops
 * Confirm neighborhood details match selected neighborhood properties in state (i.e. zipcode, town name, count of voucher holders moved, school/safety percentage, images, descriptions, links out)
 * Confirm the "Yes" checks in the `Affordable` and `Echo Benefits` cards if that neighborhood has ECC benefits
 * In "About this area section", confirm 1) images have license citation and scroll to the right if overflowing 2) Description appropriately shortened and selecting "Read more" opens full description 3) links are generated for right neighborhood zipcode (can test against prod site)
 * Toggle between the "Infos" and "Units" tab and confirm no issues displaying respective tab content
 * In "Units" content confirm neighborhood details match selected neighborhood properties in state (i.e. zipcode, town name, user's bedroom count selection, max rent for user's bedroom count selection)
 * Select one of the units site links and confirm new tab opens with correct search query params (test against same neighborhood links in prod site)
 * Navigate back to site and confirm the "units" modal is now open with functioning links to the affordability calculator and bha site (exact links to use for BHA site to be confirmed)
 * Confirm can exit this modal by clicking outside and neighborhood detail modal is still open
 * Test setting the language and confirm no issues rendering text (there are a few example translations to confirm text has changed)
 * Close modal and confirm no regressions elsewhere in app, specifically with menu modal
 * Widen screen to display desktop mode styling and confirm matches figma designs
 * Broadly compare against figma designs for any inconsistencies

Resolves #701 #663 #702

rachelekm committed:
Add initial neighborhood detail modal

rachelekm committed:
Add static Info tab content

rachelekm committed:
Issue 663: Add static map header to detail

rachelekm committed:
Add translations to neighborhood detail

rachelekm committed:
Issue 702: Add static Units tab content

rachelekm committed:
Improve modularity: pull out shared styles

rachelekm commented on src/frontend/public/locales/es/translation.json:
```diff
@@ -1,5 +1,70 @@
-    "logout": "Cerrar sesión"
+    "neighborhoodDetail": {
+        "addToFavorites": "Add to favorites",
+        "moveCountText": "¡EXAMPLE {{count}} titulares de vales ya se mudaron aquí!",
```
I added just a couple translated examples (all prefaced with EXAMPLE) just to confirm translations work when switching languages

rachelekm commented on src/frontend/src/components/CCLicensedImage.tsx:
```diff
@@ -0,0 +1,51 @@
+                    </Link>
+                </p>
+            </figcaption>
+        </figure>
```
This caption isn't in the figma designs and the existing prod site has this information in a tooltip. However, I was stuggling with populating this tooltip with dynamic content. When I could get a dynamic citation to render it would break the positioning of the tooltip and set to weird places outside the modal. I last tested this using react-aria's OverlayContainer since it uses a React Portal but it still wasn't quite working and it felt like overkill for such a small impact so I settled on figure captions.

aaronxsu replied:
Using a caption seems fine to me!

rachelekm commented on src/frontend/src/pages/NeighborhoodDetail/Map/Map.tsx:
```diff
@@ -0,0 +1,107 @@
+                padding: { top: 50, right: 50, bottom: 50, left: 50 },
+            });
+        }
+    }, [loaded]);
```
I don't love this useEffect but as far as I can tell the initialViewState doesn't take in any kind of padding option. I could calculate a buffered bounds via turf but that felt kindof heavy for the purpose vs simply waiting for the map to load and then refitting the bounds with padding. This seemed simple but does create an animated affect, I'm curious what you think.

rachelekm commented on src/frontend/src/pages/NeighborhoodDetail/index.tsx:
```diff
@@ -0,0 +1,168 @@
+import { useEffect, useState } from "react";
```
@aaronxsu and I discussed renaming this file from `index.tsx`. The verbosity is a trade-off to easier navigation in development with file name search.

aaronxsu commented on src/frontend/src/pages/NeighborhoodDetail/InfoContent.tsx:
```diff
@@ -0,0 +1,248 @@
+        display: display,
+    });
+
+    if (!neighborhood) {
```
Since `neighborhood` is of type `Neighborhood`, I wonder what this block does. Could `neighborhood` be false-y? If yes, then I'd suggest we update the type of `neighborhood` in the props of `InfoContent` component. 

Also, should the return be null or false?

rachelekm replied:
There is already a check for existing `neighborhood` in the parent component, so this type shouldn't be false-y. This block can just come out!  I scrapped it here and in `UnitsContent` as well in f9a29c6110934d92fef25ff1560a543f8118c771

aaronxsu commented on src/frontend/src/pages/NeighborhoodDetail/Map/baseMapDetailStyle.json:
```diff
@@ -0,0 +1,1630 @@
+{
```
I think it might be helpful to document somewhere how this map style differs from the map style used on the discover page.

rachelekm replied:
I added an inline comment in 932052edd0a4469ab15bf297a00f61a3f738811d

aaronxsu commented on src/frontend/src/pages/NeighborhoodDetail/Map/Map.tsx:
```diff
@@ -0,0 +1,107 @@
+            <MapContainer
+                ref={mapRef}
+                initialViewState={{
+                    bounds: bounds,
```
There seems to be a `fitBoundsOptions` prop for this `initialViewState`. Did you try it out? I wonder, if it works, then we might be able to remove the above mentioned `useEffect`

rachelekm replied:
THANK YOU! I don't know how I missed this in the docs but this is exactly what I needed and greatly simplifies this component. Incorporated `fitBoundsOptions` + clean up as part of 932052edd0a4469ab15bf297a00f61a3f738811d

aaronxsu commented on src/frontend/src/pages/NeighborhoodDetail/Map/Map.tsx:
```diff
@@ -0,0 +1,107 @@
+    const destinations = useAppSelector(selectUserDestinations);
+    const activeDestination = useAppSelector(selectActiveDestination);
+
+    let bounds = BOUNDS;
```
We might be able to wrap this block of `bounds` calculation in a `useMemo`. Any reason not using it here?

rachelekm replied:
Nice, good call. I added `useMemo` in 932052edd0a4469ab15bf297a00f61a3f738811d

aaronxsu commented on src/frontend/public/locales/en/translation.json:
```diff
@@ -1,5 +1,36 @@
     "compare": "Compare",
-    "logout": "Logout"
+    "logout": "Logout",
+    "neighborhoodDetail": {
```
Ah, I did not know you could nest translations in this way. I like it, as some certain translations only apply to specific places. This makes me what to change how I grouped the translations in https://github.com/azavea/echo-locator/pull/723

rachelekm replied:
I like it too! It makes the translation key call a little more verbose in the component but I think it's worth it for an easier time reading/updating these files

aaronxsu commented on src/frontend/src/libs/getUnitsURLByPlatform.ts:
```diff
@@ -0,0 +1,19 @@
+import type { UnitSitesKeyType } from "src/enums";
+
+const URLsBySite = {
```
This is an interesting implementation of using objects to map keys to functions. I like it.

aaronxsu commented on src/frontend/src/pages/NeighborhoodDetail/UnitsContent.tsx:
```diff
@@ -0,0 +1,363 @@
+import { Link } from "react-aria-components";
```
It is nitpicky- for the filename, I'd like to advocate that we keep using like `UnitsContent` to capitalize when it's a page or a component. Same applies to `unitsModal.txs`. What do you think?

rachelekm replied:
Makes sense to me to keep cohesive, updated in 51503f619af1e53fb86153c1c772b45bd09cdcaa

aaronxsu commented on src/frontend/src/pages/NeighborhoodDetail/unitsContent.tsx:
```diff
@@ -0,0 +1,363 @@
+                </div>
+                <div className={styles.stepGrid()}>
+                    <div
+                        className={`${sharedStyles.learnMoreLinksGroup()} ${styles.step2CardWrapper()} !bg-[#FBEBFF]`}
```
Somehow, when hovering on these clickable links, the cursor style doesn't seem to change to pointer on my end in desktop view. Is it also the case for you?

rachelekm replied:
I'll add a summarizing comment on this one but fixed in fa70f493044c02d982d3bceef870c3fc318e31eb

aaronxsu commented on src/frontend/src/pages/NeighborhoodDetail/InfoContent.tsx:
```diff
@@ -0,0 +1,280 @@
+                    </p>
+                    <div className={sharedStyles.learnMoreLinksGroup()}>
+                        <Link
+                            href={town_link}
```
Should we open these links in another tab?

rachelekm replied:
fixed in fa70f493044c02d982d3bceef870c3fc318e31eb

aaronxsu commented on src/frontend/src/components/NeighborhoodCard/ClickableNeighborhoodCard.tsx:
```diff
@@ -0,0 +1,30 @@
+    };
+
+    return (
+        <AriaButton onPress={handleOnPress}>
```
I thought using button should apply the pointer cursor style to the neighborhood card by default in the desktop view, but somehow it's not the case on my end when testing. How does it show on your end?

rachelekm replied:
I'll add a summarizing comment on this one but fixed in https://github.com/azavea/echo-locator/commit/fa70f493044c02d982d3bceef870c3fc318e31eb

aaronxsu reviewed (commented):
This is looking really great and working as advertised 👏🏼. I left some questions in the comments and would like to hear your thoughts. One other thing, as we discussed in the pairing meeting- the padding of step 1 in the units tab seems different than other steps in desktop view. Overall, this works very smoothly 👍🏼

rachelekm committed:
Simplify initial map bounds view

rachelekm committed:
Clean up link/button press styling

rachelekm committed:
Rename units files

rachelekm commented:
Thanks, @aaronxsu! This was a great review and really simplified some of the hack-y parts 🎉 

> I thought using button should apply the pointer cursor style to the neighborhood card by default in the desktop view, but somehow it's not the case on my end when testing. How does it show on your end?


Re: Some of the link/button pointer events not working as expected, I was using the react aria `Button` and `Link` out of the box since I didn't want any styling to affect the child component (like neighborhood cards or specifically-styled links). But of course that meant there was 0 styling to capture press/focus events. I created a new button variant "unstyled" for these scenarios where we want a button/button interactivity to wrap something but without any styling or sizing to affect it. I also added some of that base styling to the `learnMoreLink` style. That work is in fa70f493044c02d982d3bceef870c3fc318e31eb.

https://github.com/user-attachments/assets/3352d159-1194-4707-b525-a3e37840daa4

aaronxsu reviewed (approved):
Thanks for addressing the comments- they look great to me!

I believe one remaining detail is that the padding of step 1 in the units tab seems different than other steps in desktop view. 

After addressing the above and a changelog entry, this should be good to go. I don't need to re-review it.

ddohler commented:
Hey @rachelekm, @aaronxsu, and @philippschmitt ! 👋 Just wanted to pop in to say that I'm so happy you're getting the chance to revisit this impactful project, and the new designs look fantastic. Awesome work! 👏  I'm cheering you on from afar! 🙂

rachelekm committed:
Add styling polish

rachelekm committed:
Update CHANGELOG

rachelekm commented:
Thanks, @aaronxsu! Just addressed that final spacing issue as well as one other polish fix for the mobile-view grid in step2 of the units content. I also rebased and added a changelog entry and will squash and merge on a successful CI.
